instance/models/instance.py

```diff
@@ -268,65 +270,27 @@ def save(self, **kwargs):
-        """
-        next_server_log_entry = partial(next, server_logs, None)
-        next_instance_log_entry = partial(next, instance_logs, None)
-
-        log = []
-        instance_log_entry = next_instance_log_entry()
-        server_log_entry = next_server_log_entry()
-
-        while instance_log_entry is not None and server_log_entry is not None:
-            if server_log_entry.created < instance_log_entry.created:
-                log.append(server_log_entry)
-                server_log_entry = next_server_log_entry()
-            else:
-                log.append(instance_log_entry)
-                instance_log_entry = next_instance_log_entry()
-
-        while instance_log_entry is not None:
-            log.append(instance_log_entry)
-            instance_log_entry = next_instance_log_entry()
-
-        while server_log_entry is not None:
-            log.append(server_log_entry)
-            server_log_entry = next_server_log_entry()
-
-        return log
-
     def _get_log_entries(self, level_list=None, limit=None):
         """
         Return the list of log entry instances for the instance and its current active server,
         optionally filtering by logging level. If a limit is given, only the latest records are
         returned.
+
+        Returns oldest entries first.
         """
         # TODO: Filter out log entries for which the user doesn't have view rights
-        server_log_entry_set = self._current_server.log_entry_set
-        if level_list:
-            server_log_entry_set = server_log_entry_set.filter(level__in=level_list)
-        server_log_entry_set = server_log_entry_set.order_by('pk')
-        if limit:
-            server_log_entry_set = reversed(server_log_entry_set.reverse()[:limit])
-        else:
-            server_log_entry_set = server_log_entry_set.iterator()
-
-        instance_log_entry_set = self.log_entry_set
+        instance_type = ContentType.objects.get_for_model(self)
+        server_type = ContentType.objects.get(app_label='instance', model='openstackserver')
+        server_ids = [server.pk for server in self.server_set.all()]
+        entries = LogEntry.objects.filter(
+            (Q(content_type=instance_type) & Q(object_id=self.pk)) |
+            (Q(content_type=server_type) & Q(object_id__in=server_ids))
+        )
         if level_list:
-            instance_log_entry_set = instance_log_entry_set.filter(level__in=level_list)
-        instance_log_entry_set = instance_log_entry_set.order_by('pk')
-        if limit:
-            instance_log_entry_set = reversed(instance_log_entry_set.reverse()[:limit])
-        else:
-            instance_log_entry_set = instance_log_entry_set.iterator()
-
-        entries = self._sort_log_entries(server_log_entry_set, instance_log_entry_set)
+            entries = entries.filter(level__in=level_list)
         if limit:
-            return entries[-limit:]
-        return entries
+            return reversed(list(entries[:limit]))
```

Comment:
@bradenmacdonald Nit: When I first read this I was tempted to suggest that you could do `entries[:limit].order_by('created')` here, to be consistent with the code in the next line. But that doesn't work because "Cannot reorder a query once a slice has been taken." Maybe add a comment to that extent, to avoid potential confusion?